api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto

```diff
@@ -183,44 +183,151 @@ message JwtHeader {
+//
+// In above example, all requests (all requests matched "/" path prefix),
+// require jwt authentication from [jwt_issuer1, jwt_issuer2] and jwt_issuer2 has audiences
+// override defined.
+message RequirementRule {
+  // The route matching parameter. Only when the match is satisfied, the "requires" field will
+  // apply.
+  //
+  // For example: following match will match all requests.
+  //
+  // .. code-block:: yaml
+  //
+  //    match:
+  //       prefix: "/"
+  //
+  envoy.api.v2.route.RouteMatch match = 1 [(validate.rules).message.required = true];
+
+  // Specify list of required JwtProviders. If list is empty,  Jwt is NOT required.
+  // If any of required JwtProviders is missing, or its verification fails, the request will
+  // be rejected.
+  // If a Jwt is not required, but it is in the providers list, it will be verified.
+  // But it will not fail the request if its verification fails.
+  repeated JwtRequirement requires = 2;
+}
+
 // This is the Envoy HTTP filter config for JWT authentication.
-// [#not-implemented-hide:]
+//
+// For example:
+//
+// .. code-block:: yaml
+//
+//   providers:
+//      - issuer: issuer1
+//        audiences:
+//        - audience1
+//        - audience2
+//        remote_jwks:
+//          http_uri:
+//            uri: https://example.com/.well-known/jwks.json
+//            cluster: example_jwks_cluster
+//      - issuer: issuer2
+//        local_jwks:
+//          inline_string: jwks_string
+//
+//   rules:
+//      # Not jwt verification is required for /health path
+//      - match:
+//          prefix: "/health"
+//
+//      # Jwt verification for issuer1 is required for path prefixed with "prefix"
+//      - match:
+//          prefix: "/prefix"
+//        requires:
+//          - issuer: issuer1
+//            audiences: [audience_override]
+//
+//    # Jwt verification for issuer2 is required for all other requests.
+//      - match:
+//          prefix: "/"
+//        requires:
+//          - issuer: issuer2
+//
+//// [#not-implemented-hide:]
 message JwtAuthentication {
-  // List of JWT rules to valide.
-  repeated JwtRule rules = 1;
+  // List of JWT providers. For example:
+  //
+  // .. code-block:: yaml
+  //
+  //   providers:
+  //      - issuer: issuer1
+  //        audiences:
+  //        - audience1
+  //        - audience2
+  //        remote_jwks:
+  //          http_uri:
+  //            uri: https://example.com/.well-known/jwks.json
+  //            cluster: example_jwks_cluster
+  //      - issuer: issuer2
+  //        local_jwks:
+  //          inline_string: jwks_string
+  //
+  repeated JwtProvider providers = 1;
 
+  // DEPRECATED: Please use "routes" field. if "routes" field is not empty,
+  // this field will not be used.
   // If true, the request is allowed if JWT is missing or JWT verification fails.
   // Default is false, a request without JWT or failed JWT verification is not allowed.
   bool allow_missing_or_failed = 2;
```

Comment:
I vaguely remember someone else commenting on this field and recommending that it be removed (rather than deprecated) since this config is still in the alpha stage. Was there a final decision on that? It seems that this behavior is pretty trivial to specify as a single catch-all requirement.

Reply:
Istio is using it (as internal config between its components, not a user facing config).  So we could not remove it now,  but like to remove it once Istio code is changed to use new "rules" field.

Reply:
Got it, sounds good.